query/src/main/java/org/infinispan/query/clustered/commandworkers/CQDelete.java

```diff
@@ -27,16 +26,14 @@ CompletionStage<QueryResponse> perform(BitSet segments) {
       }
 
       SearchQueryBuilder query = queryDefinition.getSearchQueryBuilder();
-      return blockingManager.supplyBlocking(() -> fetchIds(query), this)
-            .thenApply(queryResult -> queryResult.hits().stream().map(id -> cache.remove(id) != null ? 1 : 0).reduce(0, Integer::sum))
+      return blockingManager.supplyBlocking(() -> fetchIds(query)
+                  .stream().map(id -> cache.remove(id) != null ? 1 : 0).reduce(0, Integer::sum), this)
```

Comment:
```suggestion
```suggestion
.thenCompose(queryResult -> CompletionStages.performSequentially(queryResult.hits().iterator(), cache::removeAsync,
                  Collectors.summingInt(prev -> prev != null ? 1 : 0)))
```
We should be able to keep this non blocking with the following and have the same behavior as before.
```

Reply:
@wburns, this is going to be off-topic, but do you have any brilliant ideas on how to increase concurrency? Instead of sequentially removing one by one, do it in pairs or quadruples.

Reply:
@pruivo sure, we can just change it to the performConcurrently, but I would need to add a new method that provides a Collector. It should be pretty simple. Then we can pass a Scheduler that invokes in the same thread and we have a tunable concurrency value.

Reply:
@pruivo it looks a nice idea, please open an issue for it

Reply:
Thanks @wburns I applied the idea here.

Reply:
Thanks @pruivo